src/lem/pointers.rs

```diff
@@ -7,33 +7,89 @@ use crate::{
 
 use super::Tag;
 
-/// `Ptr` is the main piece of data LEMs operate on. We can think of a pointer
-/// as a building block for trees that represent Lurk data. A pointer can be a
-/// atom that contains data encoded as an element of a `LurkField` or it can have
-/// children. For performance, the children of a pointer are stored on an
-/// `IndexSet` and the resulding index is carried by the pointer itself.
-///
-/// A pointer also has a tag, which says what kind of data it encodes. On
-/// previous implementations, the tag would be used to infer the number of
-/// children a pointer has. However, LEMs require extra flexibility because LEM
-/// hashing operations can plug any tag to the resulting pointer. Thus, the
-/// number of children have to be made explicit as the `Ptr` enum.
+/// `RawPtr` is the basic pointer type of the LEM store. An `Atom` points to a field
+/// element, and a `HashN` points to `N` children, which are also raw pointers. Thus,
+/// they are a building block for graphs that represent Lurk data.
 #[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize, Hash)]
-pub enum Ptr {
-    Atom(Tag, usize),
-    Tuple2(Tag, usize),
-    Tuple3(Tag, usize),
-    Tuple4(Tag, usize),
+pub enum RawPtr {
+    Atom(usize),
+    Hash4(usize),
+    Hash6(usize),
+    Hash8(usize),
 }
 
-impl Ptr {
-    pub fn tag(&self) -> &Tag {
+impl RawPtr {
+    #[inline]
+    pub fn is_hash(&self) -> bool {
+        matches!(
+            self,
+            RawPtr::Hash4(..) | RawPtr::Hash6(..) | RawPtr::Hash8(..)
+        )
+    }
+
+    #[inline]
+    pub fn get_atom(&self) -> Option<usize> {
+        match self {
+            RawPtr::Atom(x) => Some(*x),
+            _ => None,
+        }
```

Comment:
With [`match_opt!`](https://github.com/huitseeker/match-opt) you can turn this into 
```
match_opt!(self, RawPtr::Atom(x) => *x)
```